File: sdk/typescript/src/capabilities.ts
import { hash as blake3Js } from "blake3";
import { getPublicKey, sign as signJs, verify as verifyJs } from "@noble/ed25519";

import { getWasm } from "./wasm";

export type CapabilityToken = {
  issuer: Uint8Array;
  subject: Uint8Array;
  capability: string;
  caveats: Record<string, unknown>;
  sig: Uint8Array;
};

export function computeTokenId(token: CapabilityToken): Uint8Array {
  const payload = serializePayload(token);
  const wasm = getWasm();
  if (wasm) {
    return wasm.blake3(payload);
  }
  return toUint8Digest(blake3Js(payload));
}

export async function signToken(params: {
  issuerPrivateKey: Uint8Array;
  subject: Uint8Array;
  capability: string;
  caveats: Record<string, unknown>;
}): Promise<CapabilityToken> {
  const issuerPublicKey = await getPublicKey(params.issuerPrivateKey);
  const token: CapabilityToken = {
    issuer: issuerPublicKey,
    subject: params.subject,
    capability: params.capability,
    caveats: params.caveats,
    sig: new Uint8Array(),
  };
  const tokenId = computeTokenId(token);
  const wasm = getWasm();
  const signature = wasm
    ? await wasm.sign(tokenId, params.issuerPrivateKey)
    : await signJs(tokenId, params.issuerPrivateKey);
  return {
    issuer: issuerPublicKey,
    subject: params.subject,
    capability: params.capability,
    caveats: params.caveats,
    sig: signature,
  };
}

export async function verifyChain(tokens: CapabilityToken[], nowNs: number): Promise<void> {
  if (!tokens.length) {
    throw new Error("empty capability chain");
  }
  for (let i = 0; i < tokens.length; i += 1) {
    const token = tokens[i];
    const tokenId = computeTokenId(token);
    const wasm = getWasm();
    const ok = wasm
      ? await wasm.verify(token.sig, tokenId, token.issuer)
      : await verifyJs(token.sig, tokenId, token.issuer);
    if (!ok) {
      throw new Error("invalid capability signature");
    }
    enforceCaveats(token, nowNs);
    const next = tokens[i + 1];
    if (next && !equalBytes(token.subject, next.issuer)) {
      throw new Error("capability chain subject mismatch");
    }
  }
}

export async function enforceCapability(
  tokens: CapabilityToken[],
  required: string,
  nowNs: number
): Promise<void> {
  await verifyChain(tokens, nowNs);
  const last = tokens[tokens.length - 1];
  if (!last || last.capability !== required) {
    throw new Error("capability not granted");
  }
}

function enforceCaveats(token: CapabilityToken, nowNs: number): void {
  const notBefore = parseOptionalInt(token.caveats.not_before);
  const notAfter = parseOptionalInt(token.caveats.not_after);
  if (notBefore !== undefined && nowNs < notBefore) {
    throw new Error("capability not yet valid");
  }
  if (notAfter !== undefined && nowNs > notAfter) {
    throw new Error("capability expired");
  }
}

function parseOptionalInt(value: unknown): number | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value === "number") {
    return value;
  }
  if (typeof value === "string") {
    const parsed = Number(value);
    if (Number.isNaN(parsed)) {
      throw new Error("caveat must be int");
    }
    return parsed;
  }
  throw new Error("caveat must be int");
}

function serializePayload(token: CapabilityToken): Uint8Array {
  const payload = JSON.stringify({
    issuer: toHex(token.issuer),
    subject: toHex(token.subject),
    capability: token.capability,
    caveats: sortObject(token.caveats),
  });
  return new TextEncoder().encode(payload);
}

function toHex(data: Uint8Array): string {
  return Buffer.from(data).toString("hex");
}

function equalBytes(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

function sortObject(value: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const key of Object.keys(value).sort()) {
    const entry = value[key];
    if (entry && typeof entry === "object" && !Array.isArray(entry)) {
      out[key] = sortObject(entry as Record<string, unknown>);
    } else {
      out[key] = entry;
    }
  }
  return out;
}

function toUint8Digest(digest: Uint8Array | string): Uint8Array {
  if (typeof digest === "string") {
    return Uint8Array.from(Buffer.from(digest, "hex"));
  }
  return digest;
}

File: sdk/python/aether/capabilities.py
"""Capability token helpers for the Python SDK."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Iterable, Mapping

from blake3 import blake3

from .crypto import sign as sign_event_id
from .crypto import verify as verify_event_id


@dataclass(frozen=True)
class CapabilityToken:
    issuer: bytes
    subject: bytes
    capability: str
    caveats: Mapping[str, object]
    sig: bytes


def compute_token_id(token: CapabilityToken) -> bytes:
    payload = _serialize_payload(token)
    return blake3(payload).digest()


def sign_token(
    *, issuer_private_key: bytes, subject: bytes, capability: str, caveats: Mapping[str, object]
) -> CapabilityToken:
    token = CapabilityToken(
        issuer=_parse_hex_or_bytes(issuer_private_key, "issuer", 32),
        subject=_parse_hex_or_bytes(subject, "subject", 32),
        capability=capability,
        caveats=caveats,
        sig=b"",
    )
    token_id = compute_token_id(token)
    sig = sign_event_id(token_id, issuer_private_key)
    return CapabilityToken(
        issuer=token.issuer,
        subject=token.subject,
        capability=token.capability,
        caveats=token.caveats,
        sig=sig,
    )


def verify_chain(
    tokens: Iterable[CapabilityToken],
    *,
    now_ns: int,
    usage: Mapping[bytes, int] | None = None,
) -> None:
    chain = list(tokens)
    if not chain:
        raise ValueError("empty capability chain")

    for idx, token in enumerate(chain):
        token_id = compute_token_id(token)
        if not verify_event_id(token_id, token.sig, token.issuer):
            raise ValueError("invalid capability signature")

        _enforce_caveats(token, now_ns, usage or {})

        if idx + 1 < len(chain):
            next_token = chain[idx + 1]
            if token.subject != next_token.issuer:
                raise ValueError("capability chain subject mismatch")


def _enforce_caveats(
    token: CapabilityToken,
    now_ns: int,
    usage: Mapping[bytes, int],
) -> None:
    caveats = token.caveats or {}
    not_before = _parse_optional_int(caveats.get("not_before"))
    not_after = _parse_optional_int(caveats.get("not_after"))
    max_uses = _parse_optional_int(caveats.get("max_uses"))

    if not_before is not None and now_ns < not_before:
        raise ValueError("capability not yet valid")
    if not_after is not None and now_ns > not_after:
        raise ValueError("capability expired")
    if max_uses is not None:
        token_id = compute_token_id(token)
        if usage.get(token_id, 0) >= max_uses:
            raise ValueError("capability usage exceeded")


def _serialize_payload(token: CapabilityToken) -> bytes:
    payload = {
        "issuer": token.issuer.hex(),
        "subject": token.subject.hex(),
        "capability": token.capability,
        "caveats": token.caveats,
    }
    return json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")


def _parse_optional_int(value: object) -> int | None:
    if value is None:
        return None
    if isinstance(value, bool):
        raise ValueError("caveat must be int")
    if isinstance(value, int):
        return value
    if isinstance(value, str):
        try:
            return int(value)
        except ValueError as exc:
            raise ValueError("caveat must be int") from exc
    raise ValueError("caveat must be int")


def _parse_hex_or_bytes(value: object, field: str, size: int) -> bytes:
    if isinstance(value, bytes):
        data = value
    elif isinstance(value, str):
        try:
            data = bytes.fromhex(value)
        except ValueError as exc:
            raise ValueError(f"{field} must be hex") from exc
    else:
        raise ValueError(f"{field} must be bytes or hex string")

    if len(data) != size:
        raise ValueError(f"{field} must be {size} bytes")
    return data
